### DISCUSSION TITLE
Getting Band from MSM messages?

### DISCUSSION THREAD

**Discussion by gytisgreitai** [semuconsulting/pyrtcm#43] [Q&A] [answered]
Hey,

so I can resolve signal ID by doing `cell2prn` on a RTCMMessage . But how do I get the Band (e.g. in case of 1097 for Galileo that would be E1/E5 etc.) ? Am I missing something obvious, or `rtcmtables.py` does not have that information, and my best bet would be to create a map from rinex304 which would map the signal tables `*_SIG_MAP` to bands ?

Thanks!

**Comment by semuadmin** [comment, marked as answer]
Hi @gytisgreitai 

You're on the right track - provided you have the keyword argument `labelmsm` set to true (the default), pyrtcm will parse relevant MSM message attributes with a label signifying both the PRN and the signal RINEX  code - see https://github.com/semuconsulting/pyrtcm/blob/main/src/pyrtcm/rtcmtables.py. Each signal RINEX code corresponds to a frequency band and signal channel.

So, for example, a value of `DF405_10(014,5I)` signifies PRN 14 and a signal RINEX code of `5I`, which corresponds to the frequency band/signal`E5a/I`. A value of `DF405_10(023,1A)` signifies PRN 23 and a signal RINEX code of `1A` which corresponds to frequency band/signal `E1/A`. The frequency-signal ID mappings for each constellation are given in the proprietary RTCM3 specification (_which I can't reproduce in full here for copyright reasons_), but this information can be found in public domain RINEX documentation, e.g. from RINEX 3.04 section 5.1 observation codes:

<img width="806" alt="Screenshot 2023-10-01 at 20 06 54" src="https://github.com/semuconsulting/pyrtcm/assets/28569967/01177cd0-0b6d-44f2-a2be-19a3b13e28d8">

Hope this helps.

**Comment by gytisgreitai** [comment]
Hey, thanks for detailed answer! 

> The frequency-signal ID mappings for each constellation are given in the proprietary RTCM3 specification (which I can't reproduce in full here for copyright reasons)

So what you are saying is that one can't put something like this in their github repo?
```
GPS_SIGAL_TO_BAND = {
  'rinex_code': 'band',
 <...>
}

GLONASS_SIGAL_TO_BAND = {
  'rinex_code': 'band',
 <...>
}

etc.
```
Cause I've mapped the codes from the docs, and wrote a simple python script to print the GNSS system + satellite + bands I'm seeing, e.g.:
```
GPS 2(L1,L2) 3(L1,L2,L5) 4(L1,L2,L5) 6(L1,L2,L5) 9(L1,L2,L5) 17(L1,L2) 19(L1,L2) 21(L1,L2) 31(L1,L2) 32(L1,L2,L5)
GALILEO 7(E1,E5b,E5,E5a) 10(E1,E5b,E5,E5a) 12(E1,E5b,E5,E5a) 19(E1,E5b,E5,E5a) 26(E1,E5b,E5,E5a) 33(E1,E5b,E5,E5a)
GLONASS 1(G1,G2) 2(G1,G2) 3(G1,G2) 11(G1,G2) 12(G1,G2) 17(G1,G2) 18(G1,G2)
```

**Reply by semuadmin** [reply]
Provided the information is in the public domain, there's nothing stopping it being incorporated into a public GitHub repo like pyrtcm - I just can't upload the RTCM3 specification itself here.

My `labelmsm` function uses the the signal RINEX ID for brevity, as it contains all the information most end users would need, but if you fancy submitting a PR to add this additional lookup, be my guest!

**Comment by semuadmin** [comment]
@gytisgreitai 

FYI I've just uploaded an enhancement in [RC 1.0.13](https://github.com/semuconsulting/pyrtcm/tree/RC-1.0.13) which hopefully addresses your needs.

The `cell2prn()` method now includes an additional optional parameter `sigcode`, with values of: 
- 0 = use frequency band
- 1 = use signal RINEX code (the current default behaviour)

The `labelmsm` keyword argument in `RTCMReader` and `RTCMMessage` has been updated accordingly:
- 0 = no MSM label
- 1 = label with PRN & signal RINEX code (the current default behaviour)
- 2 = label with PRN and frequency band

**Reply by gytisgreitai** [reply]
oh wow, that was fast, and works nicely. Thank you very much for a great library!
